### PR TITLE
DISCO-240 Added BodyPortal

### PR DESCRIPTION
**Pull request by Dantemss**
Renders NavBar, Modals, Overlay, Toasts outside of root so we can use `<main>` as our React root.
`BodyPortalSlotsContext` can be used in the layout to specify the ordering of the portals and the root.

This seems to work, though I wonder if we want ErrorModal to have a different slot than Modal, for example.
If we don't do that, then the ordering between Modal and ErrorModal might change when re-rendering.

### CODE REVIEW

File: src/components/BodyPortal.tsx [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import { createPortal } from 'react-dom';
+import { BodyPortalSlotsContext } from './BodyPortalSlotsContext';
+
+const getInsertBeforeTarget = (bodyPortalSlots: string[], slot: string) => {
+  // Note: If the slot is not found in bodyPortalSlots, this code will insert the tag before the first slot
+  //       This is not recommended usage, as the ordering will then depend on the rendering order and may change
+  for (let index = bodyPortalSlots.findIndex((sl) => sl === slot) + 1; index < bodyPortalSlots.length; index++) {
+    const sl = bodyPortalSlots[index];
+    const tag = sl === 'main'
+    ? document.body.querySelector('main')
```
Comment (Dantemss) [inline src/components/BodyPortal.tsx:11]: Another option here is to find main by `id="root"` which mirrors what React does more closely, though I find `main` works better as the slot name... I'm not sure which is best.

**Comment by Dantemss** [comment]
Updating `@types/styled-components` is a pain because it updates `@types/hoist-non-react-statics` which tries to pull in React 18 types.

**Review by jivey — CHANGES_REQUESTED** [review]
One of the things I wasn't sure about is if this would break the ability to use the components solo, but it seems to work when I preview them in ladle without the wrapper (add `mode=preview` to the URL or click the expand icon.) But it doesn't render with the normal ladle view, could we make that work by changing the ladle wrapper?

**Comment by Dantemss** [comment]
> One of the things I wasn't sure about is if this would break the ability to use the components solo, but it seems to work when I preview them in ladle without the wrapper (add `mode=preview` to the URL or click the expand icon.) But it doesn't render with the normal ladle view, could we make that work by changing the ladle wrapper?

I guess they get pushed off-screen by the ladle stuff... I'll try to set something up to prevent this.

**Comment by Dantemss** [comment]
@jivey well, I couldn't figure out how to change the ladle stuff but I fixed them with some custom CSS...

**Review by jivey — APPROVED** [review]
The portal stuff looks alright to me, might be good to have @TomWoodward look at it too

File: src/components/BodyPortal.tsx [hunk 2]
```diff
@@ -0,0 +1,56 @@
+import React from 'react';
+import { createPortal } from 'react-dom';
+import { BodyPortalSlotsContext } from './BodyPortalSlotsContext';
+
+const getInsertBeforeTarget = (bodyPortalSlots: string[], slot: string) => {
+  // Note: If the slot is not found in bodyPortalSlots, this code will insert the tag before the first slot
```
Comment (TomWoodward) [inline src/components/BodyPortal.tsx:6]: there are situations, such as rendering an absolutely positioned dialog, when we might not care to specify the slot order and its ok for it to get added wherever. i think typically it would be added at the bottom of the body, but i'm not sure it'll really matter if it goes at the top instead

File: src/components/NavBar.stories.tsx [hunk 3]
```diff
@@ -29,24 +29,30 @@ const StyledWrapper = styled.div`
   height: 100%
 `;
 
+const StyledNavBar = styled(NavBar)`
+  position: fixed;
+  left: 2rem;
+  top: 2rem;
+  width: calc(100% - 36rem);
+`;
```
Comment (TomWoodward) [inline src/components/NavBar.stories.tsx:37]: why was it necessary to add styling for this that wasn't there already?

File: src/components/ToastContainer.stories.tsx [hunk 4]
```diff
@@ -1,9 +1,16 @@
+import styled from 'styled-components';
 import { ToastData } from '../../src/types';
 import { ToastContainer } from './ToastContainer';
 
+const StyledToastContainer = styled(ToastContainer)`
+  top: 2rem;
+  left: 2rem;
+  right: unset;
+`;
```
Comment (TomWoodward) [inline src/components/ToastContainer.stories.tsx:9]: again, why did you need to add these styles?

File: src/components/NavBar.stories.tsx (hunk 3 above)
Comment (Dantemss) [reply src/components/NavBar.stories.tsx:37]: I couldn't get ladle to make a layout that made sense. They have `main` a few layers in and there's no way to change that, or to give it an id of "root".

File: src/components/BodyPortal.tsx [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+    if (tag) { return tag; }
+  }
+
+  // None of the slots after this one are present in the DOM, so just append it instead
+  return null;
+}
+
+export const BodyPortal = ({
+  children, className, role, slot, tagName
+}: React.PropsWithChildren<{className?: string; role?: string; slot: string; tagName?: string}>) => {
```
[the remark this answers is not in the document]
Comment (Dantemss) [reply src/components/BodyPortal.tsx]: Changed slot to be optional

File: src/components/BodyPortal.tsx (hunk 2 above)
Comment (Dantemss) [reply src/components/BodyPortal.tsx]: Changed getInsertBeforeTarget to append by default if the slot is not found

**Review by TomWoodward — APPROVED** [review]
